docs/asdf/developer_versioning.rst

```diff
@@ -0,0 +1,253 @@
+to the ASDF Standard version in use, which dictates the subset of
+schema versions that are available.  From the subset of AsdfType
+subclasses that handle those schema versions, it selects the subclass
+that is able to handle the type of the core object being serialized.
+
+The rules for selecting an AsdfType for a given serializable object
+are implemented by ``asdf.type_index.AsdfTypeIndex.from_custom_type``.
+
+Implementing updates to the standard
+------------------------------------
+
+Let's assume that there is a new standard version, 2.0.0, which
+includes one entirely new core schema, ``core/new_object-1.0.0.yaml``,
+one backwards-compatible update to an existing schema,
+``core/updated_object-1.1.0.yaml``, and one breaking change to an
+existing schema, ``core/breaking_object-2.0.0.yaml``.  The following
+sections walk through the steps we'll need to take to support
+this new material.
+
+Update the asdf-standard submodule commit pointer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The asdf-standard repository is integrated into the asdf repository
+as a submodule.  To pull in new commits from the remote master (
+assumed to be named ``origin``:
+
+.. code-block:: console
+
+    $ cd asdf-standard
+    $ git fetch origin
+    $ git checkout origin/master
+
+Support the new standard version
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The list can be found in ``asdf.versioning.supported_versions``.
+Add ``AsdfVersion("2.0.0")`` to the end of the list
+(maintaining the sort order).  This new version will become the default
+for new files, but we can update the definition of
+``asdf.versioning.default_version`` if that is undesirable.
+
+Support the new schema
+~~~~~~~~~~~~~~~~~~~~~~
+
+Schemas for previously unsupported objects are straightforward, since
+we don't need to worry about compatibility issues.  Create a new
+AsdfType subclass with ``name`` and ``version`` set appropriately:
+
+.. code-block:: python
+
+    class NewObjectType(AsdfType):
+        name = "core/new_object"
+        version = "1.0.0"
+
+In a real-life scenario, we'd need to actually support (de)serialization
+in some way, but those details are beyond the scope of this document.
+
+Support the backwards-compatible schema
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since our updated_object-1.1.0.yaml schema is backwards-compatible,
+we can share the same AsdfType subclass between it and the previous
+version.  Presumably there exists an AsdfType that looks something
+like this:
+
+.. code-block:: python
+
+    class UpdatedObjectType(AsdfType):
+        name = "core/updated_object"
+        version = "1.0.0"
+
+We'll need to update the version, and list 1.0.0 as a supported
+version, so that this class can continue to handle it:
+
+.. code-block:: python
+
+    class UpdatedObjectType(AsdfType):
+        name = "core/updated_object"
+        version = "1.1.0"
+        supported_versions = {"1.0.0", "1.1.0"}
+
+Support the breaking schema
```

Comment:
I wonder if we shouldn't add a statement suggesting that backward incompatible changes should be highly discouraged. I can't say that they should never be permitted (I suspect we can think of examples), but sometimes a new type object (and core tag name) may be much more preferable than doing this. I guess the issue comes down to when we are creating a Python object that we do not want to change the class name, but do want to change how it is serialized in a backward incompatible way. Since the class type is used to find the AsdfType, does he machinery support selecting between two different AsdfTypes depending on some indication of which now is most recent when writing? Does what I'm saying make any sense? :-)

Reply:
Happily, it does support selecting between different AsdfTypes!  The first thing the library does on write is consult the version map for relevant schema versions, and then select the subset of AsdfTypes that support those versions.  So we can have two AsdfTypes that (de)serialize the same type of object, and as long as they specify different schema versions, they won't conflict with one another.